rfc/rfc-43/rfc-43.md

```diff
@@ -0,0 +1,222 @@
+
+## Abstract
+
+Hudi table needs compaction/clustering to rewrite data. Currently, schedule compaction/clustering job provides Three ways:
+
+- Inline, execute compaction/clustering job and writing job in the same application, perform the compaction/clustering job and writing job serially.
+
+- Async, execute compaction/clustering job and writing job in the same application, Async parallel execution of compaction/clustering job and write job.
+
+- Independent compaction/clustering job, execute an async compaction/clustering job of another application.
+
+With the increase in the number of HUDI tables, due to a lack of management capabilities, maintenance costs will become higher. This proposal is to implement an independent compaction/clustering Service to manage the Hudi compaction/clustering job.
+
+## Background
+
+In the current implementation, if the HUDI table needs do compact/cluster, it only has three ways:
+
+1. Use inline compaction/clustering, in this mode the job will be block writing job.
+
+2. Using Async compaction/clustering, in this mode the job execute async but also sharing the resource with HUDI to write a job that may affect the stability of job writing, which is not what the user wants to see.
+
+3. Using independent compaction/clustering job is a better way to schedule the job, in this mode the job execute async and do not sharing resources with writing job, but also has some questions:
+   1. The user needs to manually start an async compaction/clustering application, which means that the user needs to maintain two jobs.
+   2. With the increase in the number of HUDI jobs, there is no unified service to manage compaction/clustering jobs (monitor, retry, history, etc...), which will make maintenance costs increase.
+
+With this effort, we want to provide an independent compaction/clustering Service, it will have these abilities:
+
+- Provides a pluggable execution interface that can adapt to multiple execution engines, such as Spark and Flink.
+
+- With the ability to failover, need to be persisted compaction/clustering message.
+
+- Perfect metrics and reuse HoodieMetric expose to the outside.
+
+- Provide automatic failure retry for compaction/clustering job.
+
+## Implementation
+
+![](service.png)
+
+### Client
+
+Register to Service with CompactionConfig/ClusteringConfig when the job starts. When generating a compaction/clustering plan, report to the compaction/clustering Service address specified in HoodieConfig and request the Service to stop the corresponding compaction/clustering job when the Service is not required or rollback compaction/clustering instant.
+
+### Request Handler
+
+Receive the client's request and save the plan in the meta table.
+
+### Meta Table
+
+Meta table has been implemented by using an internal HUDI MOR Table to store the required metadata for the compaction/clustering message. This table will be internal to a dataset and will not be exposed directly to the user to write/modify.
```

Comment:
Using metadata table for this has some issues:
1. Scale - Since each write to metadata table will be a deltacommit which may take many seconds (my tests should 5 to 10seconds is minimum), this restricts us to maximum number of tables possible
2. Job state tracking (start job, complete job, job error) means further writes into the metadata table. This will further lower the number of possible requests / sec that can be served.

I feel there are better approached for saving this information:
1. Using rocksdb 
2. Using simple "job files" in yaml format in the folder ".hoodie/service". Each job scheduling writes a new job file and on completion we can simply delete the file or archive it. This will be more work than rocksdb.
3.  Using an external store like MySQL or MongoDB (downside is external depedecy ofcourse).

We have a scale of 10000+ tables with ingestion running atleast every 30 mins.

In any case, this could be an Interface with multiple implementations.

Reply:
Thanks for review.
I will update this document later, the current implementation is already an Interface with multiple implementations.

Reply:
This is a great idea, we can use a built-in rocksdb by default, considering that the introduction of external dependencies will increase the complexity of the system to a certain extent.